python/cudf/cudf/tests/test_df_protocol.py

```diff
@@ -232,3 +238,13 @@ def test_NA_mixed_dtype():
         string=[None, None, None, "df protocol", None],
     )
     assert_df_unique_dtype_cols(data_mixed)
+
+
+@pytest.mark.skipif(
+    not PANDAS_GE_150,
+    reason="Pandas versions < 1.5.0 do not support interchange protocol",
+)
+def test_from_cpu_df(pandas_df):
+    df = pd.DataFrame({"a": [1, 2, 3], "b": ["x", "y", "z"]})
+    with pytest.raises(NotImplementedError):
+        cudf.from_dataframe(df, allow_copy=True)
```

Comment:
Can we assert this actually worked?

Reply:
The goal of this PR was to raise the correct error message when consuming CPU-backed DataFrames. After reading your comment, I figured we might as well implement the bits needed to actually consume them. In the process I discovered a slew of issues that I've tried to address in this PR. I've re-requested your review, but I'll leave comments inline where I think it would be useful.